# ==============================================================================
# REGRESSOR COMMIT DIFF
# ==============================================================================
# File: layout/printing/nsPrintJob.cpp
# Commit: dfbaf61ce16a
# Full Hash: dfbaf61ce16a579fb9a2251feabd524dd2f1cdc3
# Author: Jonathan Watt <jwatt@jwatt.org>
# Date: 2020-07-11 21:52:53
# Regressor Bug: 1557983
# File Overlap Count: 1
# Description:
#   Bug 1557983. Significantly simplify and fix the 'beforeprint'/'afterprint' dispatching code.
#   
#   This fixes two issue.
#   
#   First, the code shouldn't be dispatching these events every time it gets a new
# ==============================================================================

diff -r e3325d8cbd90 -r dfbaf61ce16a layout/printing/nsPrintJob.cpp
--- a/layout/printing/nsPrintJob.cpp	Sat Jul 11 12:11:20 2020 +0300
+++ b/layout/printing/nsPrintJob.cpp	Sat Jul 11 09:06:25 2020 +0000
@@ -168,6 +168,26 @@
 #  define DUMP_DOC_TREELAYOUT
 #endif
 
+static CallState CollectDocuments(Document& aDoc,
+                                  nsTArray<nsCOMPtr<Document>>& aDocs) {
+  aDocs.AppendElement(&aDoc);
+  auto recurse = [&aDocs](Document& aSubDoc) {
+    return CollectDocuments(aSubDoc, aDocs);
+  };
+  aDoc.EnumerateSubDocuments(recurse);
+  return CallState::Continue;
+}
+
+static void DispatchEventToWindowTree(Document& aDoc, const nsAString& aEvent) {
+  nsTArray<nsCOMPtr<Document>> targets;
+  CollectDocuments(aDoc, targets);
+  for (nsCOMPtr<Document>& doc : targets) {
+    nsContentUtils::DispatchTrustedEvent(doc, doc->GetWindow(), aEvent,
+                                         CanBubble::eNo, Cancelable::eNo,
+                                         nullptr);
+  }
+}
+
 class nsScriptSuppressor {
  public:
   explicit nsScriptSuppressor(nsPrintJob* aPrintJob)
@@ -664,6 +684,14 @@
   NS_ENSURE_SUCCESS(rv, rv);
 
   {
+    if (!aSourceDoc->IsStaticDocument()) {
+      // This is the original document.  We must send 'beforeprint' and
+      // 'afterprint' events to give the document the chance to make changes
+      // for print output.  (Obviously if `aSourceDoc` is a clone, it already
+      // has these changes.)
+      DispatchEventToWindowTree(*aSourceDoc, u"beforeprint"_ns);
+    }
+
     nsAutoScriptBlocker scriptBlocker;
     printData->mPrintObject = MakeUnique<nsPrintObject>();
     rv = printData->mPrintObject->InitAsRootObject(docShell, aSourceDoc,
@@ -678,6 +706,10 @@
 
     BuildNestedPrintObjects(printData->mPrintObject->mDocument,
                             printData->mPrintObject, &printData->mPrintDocList);
+
+    if (!aSourceDoc->IsStaticDocument()) {
+      DispatchEventToWindowTree(*aSourceDoc, u"afterprint"_ns);
+    }
   }
 
   // The nsAutoScriptBlocker above will now have been destroyed, which may
